tsweb/jsonhandler.go

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsweb
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"tailscale.com/types/logger"
+)
+
+// JSONHandler is like net/http.Handler, but the handler has JSON and Form
+// data parsed for it out of http.Request and returns JSON or an error
+// that is converted to JSON instead of writing to its ResponseWriter.
+type JSONHandler interface {
+	// ServeHTTPJSON is like http.Handler.ServeHTTP,
+	// except optimized for JSON REST APIs.
+	//
+	// If the request is a POST, then the request data is decoded
+	// and passed as a map to ServeHTTPJSON. The data can be either
+	// a url-encoded form or a JSON blob. The JSON blob is required
+	// to be a property set (that is, start with '{').
+	//
+	// The caller will write a response to the ResponseWriter.
+	// The response is always a JSON blob, either:
+	//
+	//	{"success": true, "data": <encoded obj>}
+	//
+	// or
+	//
+	//	{"success": false, "error": "<error message>"}
+	//
+	// If ServeHTTPJSON returns no error, the returned obj is
+	// encoded as a JSON property list and sent to the client.
+	//
+	// If ServeHTTPJSON returns an error, it caller should handle
+	// an error by serving an HTTP 500 response to the user. The
+	// error details must not be sent to the client, as they may
+	// contain sensitive information. If the error is an
+	// HTTPError, though, callers should use the HTTP response
+	// code and message as the response to the client.
+	ServeHTTPJSON(http.ResponseWriter, *http.Request, map[string]interface{}) (obj interface{}, err error)
```

Comment:
I'd name that map input argument something.

But the new type and type unsafety is a bit gross. I'm leaning towards something like this instead:

```go
    package tsweb

    // NewJSONHandler wraps an HTTP handler function with a version that automatically
    // unmarshals and marshals requests and responses respectively into fn's arguments
    // and results.
    //
    // fn must have a signature with 2 or 3 input arguments. The first two must be http.ResponseWriter and
    // *http.Request. The optional third argument can be of any type.
    // fn's result parameter can be either (error) or (T, error), where T is the JSON-marshalled result type.
    func NewJSONHandler(fn interface{}) http.Handler {
           // reflect.Call games
    }
```

Then all the type unsafety is in one place and we don't introduce a new type.

(Admittely the signature is a bit runtime.SetFinalizer-y).

We can panic at runtime if the signature is bogus which will help with testing or just generate 500s.

Reply:
Yeah I like that better. Will work on it slowly over the next couple of days.